src/rgw/rgw_admin.cc

```diff
@@ -6186,23 +6186,30 @@ int main(int argc, const char **argv)
     formatter->open_object_section("result");
     formatter->dump_string("bucket", bucket_name);
     formatter->open_array_section("objects");
-    uint16_t attempt = 0;
+
+    constexpr uint32_t NUM_ENTRIES = 1000;
+    uint16_t attempt = 1;
     while (is_truncated) {
       RGWRados::ent_map_t result;
       int r =
 	store->getRados()->cls_bucket_list_ordered(bucket_info, RGW_NO_SHARD, marker,
-				       prefix, 1000, true,
+				       prefix, NUM_ENTRIES, true,
+				       attempt,
 				       result, &is_truncated, &marker,
                                        null_yield,
-				       ++attempt,
 				       bucket_object_check_filter);
 
       if (r < 0 && r != -ENOENT) {
         cerr << "ERROR: failed operation r=" << r << std::endl;
+      } else if (r == -ENOENT) {
+        break;
       }
 
-      if (r == -ENOENT)
-        break;
+      if (result.size() < NUM_ENTRIES / 8) {
+	++attempt;
+      } else if (result.size() > NUM_ENTRIES * 7 / 8 && attempt > 1) {
+	--attempt;
```

Comment:
--attempt?

Reply:
When I tested it and created on the order of 3000+ incomplete multipart uploads and then a few files with 97 shards. The calculated per-shard request is 11. And even without the use of "attempt" to allow the request count to grow exponentially, it just kept moving through the listing 11 at a time until it completed the listing. So I think it's robust to this type of error.

The decrement of attempt is protected by `attempt > 1`, so it should always be positive.  However when I wrote it for the regular listing, it literally counted the number of attempts. The uses in each of rgw_admin.cc and rgw_bucket.cc are not the same. So I'd prefer a more general term than "attempt".